--- a/arch/arm/dts/rockchip-u-boot.dtsi
+++ b/arch/arm/dts/rockchip-u-boot.dtsi
@@ -99,19 +99,16 @@
 					};
 #endif
 				};
-				@tee-SEQ {
-					fit,operation = "split-elf";
-					description = "TEE";
+				op-tee {
+					description = "OP-TEE";
 					type = "tee";
 					arch = "arm64";
-					os = "tee";
+					os = "op-tee";
 					compression = "none";
-					fit,load;
-					fit,entry;
-					fit,data;
+					load = <0x30000000>;
+					entry = <0x30000000>;
 
 					tee-os {
-						optional;
 					};
 #ifdef CONFIG_SPL_FIT_SIGNATURE
 					hash {
@@ -157,7 +154,7 @@
 					description = "NAME.dtb";
 					fdt = "fdt-SEQ";
 #ifdef CONFIG_ARM64
-					fit,firmware = "atf-1", "u-boot";
+					fit,firmware = "atf-1", "u-boot", "op-tee";
 #else
 					fit,firmware = "op-tee", "u-boot";
 #endif
