diff --git a/arch/arm/dts/rockchip-u-boot.dtsi b/arch/arm/dts/rockchip-u-boot.dtsi
<<<<<<< Updated upstream
index c8c928c7e5..a2d27bb884 100644
=======
index c8c928c7e5..13e69f7cae 100644
>>>>>>> Stashed changes
--- a/arch/arm/dts/rockchip-u-boot.dtsi
+++ b/arch/arm/dts/rockchip-u-boot.dtsi
@@ -99,19 +99,16 @@
 					};
 #endif
 				};
-				@tee-SEQ {
-					fit,operation = "split-elf";
-					description = "TEE";
+				op-tee {
+					description = "OP-TEE";
 					type = "tee";
 					arch = "arm64";
-					os = "tee";
+					os = "op-tee";
 					compression = "none";
-					fit,load;
-					fit,entry;
-					fit,data;
+					load = <0x30000000>;
+					entry = <0x30000000>;
 
 					tee-os {
-						optional;
 					};
 #ifdef CONFIG_SPL_FIT_SIGNATURE
 					hash {
<<<<<<< Updated upstream
=======
@@ -161,7 +158,7 @@
 #else
 					fit,firmware = "op-tee", "u-boot";
 #endif
-					fit,loadables;
+					fit,loadables = "aft-2", "op-tee";
 				};
 			};
 		};
>>>>>>> Stashed changes
